python/packages/autogen-core/src/autogen_core/models/_cache.py

```diff
@@ -0,0 +1,197 @@
+import hashlib
+import json
+import warnings
+from typing import Any, AsyncGenerator, List, Mapping, Optional, Sequence, Union, cast
+
+from .._cache_store import CacheStore
+from .._cancellation_token import CancellationToken
+from ..tools import Tool, ToolSchema
+from ._model_client import (
+    ChatCompletionClient,
+    ModelCapabilities,  # type: ignore
+    ModelInfo,
+)
+from ._types import (
+    CreateResult,
+    LLMMessage,
+    RequestUsage,
+)
+
+CHAT_CACHE_VALUE_TYPE = Union[CreateResult, List[Union[str, CreateResult]]]
+
+
+class ChatCompletionCache(ChatCompletionClient):
+    """
+    A wrapper around a ChatCompletionClient that caches creation results from an underlying client.
+    Cache hits do not contribute to token usage of the original client.
+    """
+
+    def __init__(self, client: ChatCompletionClient, store: CacheStore[CHAT_CACHE_VALUE_TYPE]):
+        """
+        Initialize a new ChatCompletionCache.
+
+        First initialize (for eg) a Redis store:
+
+        ```python
```

Comment:
Use:

```
.. code-bloc:: python

```

For the Python code. We will need to add `redis` and `diskcache` to the doc dependencies to build the doc, as we type check all the code blocks included in the API docs.